### PR TITLE
KAFKA-14780: Fix flaky test 'testSecondaryRefreshAfterElapsedDelay'

### PR DESCRIPTION
**Pull request by olalamichelle**
"The test RefreshingHttpsJwksTest#testSecondaryRefreshAfterElapsedDelay relies on the actual system clock, which makes it frequently fail. The fix adds a second constructor that allows for passing a ScheduledExecutorService to manually execute the scheduled tasks before refreshing. The fixed task is much more robust and stable.The original JIRA ticket can be found here: [https://issues.apache.org/jira/browse/KAFKA-14780?jql=project%20%3D%20KAFKA%20AND%20status%20%3D%20Open%20AND%20text%20~%20%22RefreshingHttpsJwksTest%22%20ORDER%20BY%20created%20DESC](url)

### CODE REVIEW

**Comment by olalamichelle** [comment]
@divijvaidya Sorry It is my first time trying to contribute to Kafka. I didn't see any error log with the build failure. Could you help review the PR and give me some guidance on how to rerun the tests to deflaky it? Or do you see any error log with the failure?

**Comment by divijvaidya** [comment]
Thank you for your first contribution! 

If we look at metrics for last 28 days, this test has been flaky 1% (10 out of 1300 builds) of the time as per https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&search.timeZoneId=Europe/Berlin&tests.container=org.apache.kafka.common.security.oauthbearer.internals.secured.RefreshingHttpsJwksTest and it passes 99% of the time. Hence, it would be difficult to reproduce the flaky behaviour for this one.

You can see the error logs at https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&search.timeZoneId=Europe/Berlin&tests.container=org.apache.kafka.common.security.oauthbearer.internals.secured.RefreshingHttpsJwksTest&tests.test=testSecondaryRefreshAfterElapsedDelay() 

To re-run the test and trying to reproduce flakiness, I would suggest to use this command locally on your machine: https://github.com/apache/kafka/blob/trunk/README.md#repeatedly-running-a-particular-unitintegration-test

**Comment by olalamichelle** [comment]
Hi Divij,

Thank you for your reply. I mean the failure of the existing CI job
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14078/1/pipeline.
It seems not to be caused by my change? And how can I rerun the CI? For the
flakiness of RefreshingHttpsJwksTest I think my PR already resolved it.

Regards and thanks,

Fei

**Comment by divijvaidya** [comment]
Hi Fei

CI tests are re-triggered on new commits or when a committer restarts them. Those are the only two options right now.
Regarding other flaky tests, if you run `./gradlew unitTest` and `./gradlew integrationTest` locally, they should pass. You can mention in the description that these tests pass locally for you and after that the person who will merge can determine if the flaky tests in CI are related to your change or not prior to merging. You can also look at JIRA tickets and usually the existing flaky tests have an open JIRA ticket.

**Comment by divijvaidya** [comment]
I have just re-triggered CI tests for this PR at https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14078/2/pipeline

**Comment by olalamichelle** [comment]
Hi Divij,

It is stating that "all checks have failed and this commit has test failures". But upon checking the merge details, I was not able to find the exact failure info. Could you please share some advice? Thank you so much!

**Comment by showuon** [comment]
The failed tests are listed here: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14078/2/
They are all unrelated to your change. I'll review this PR later. Thanks for the contribution.

**Comment by divijvaidya** [comment]
Unrelated failures:
```
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testSyncTopicConfigs()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14078/2/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_11_and_Scala_2_13___testSyncTopicConfigs__/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14078/2/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationExactlyOnceTest/Build___JDK_17_and_Scala_2_13___testOffsetTranslationBehindReplicationFlow__/)
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14078/2/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationExactlyOnceTest/Build___JDK_20_and_Scala_2_13___testOffsetTranslationBehindReplicationFlow__/)
[Build / JDK 20 and Scala 2.13 / kafka.admin.TopicCommandIntegrationTest.testListTopicsWithExcludeInternal(String).quorum=zk](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14078/2/testReport/junit/kafka.admin/TopicCommandIntegrationTest/Build___JDK_20_and_Scala_2_13___testListTopicsWithExcludeInternal_String__quorum_zk/)
[Build / JDK 20 and Scala 2.13 / kafka.admin.TopicCommandIntegrationTest.testCreateWithConfigs(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14078/2/testReport/junit/kafka.admin/TopicCommandIntegrationTest/Build___JDK_20_and_Scala_2_13___testCreateWithConfigs_String__quorum_kraft/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testSyncTopicConfigs()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14078/2/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_8_and_Scala_2_12___testSyncTopicConfigs__/)
[Build / JDK 8 and Scala 2.12 / kafka.api.TransactionsTest.testBumpTransactionalEpoch(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14078/2/testReport/junit/kafka.api/TransactionsTest/Build___JDK_8_and_Scala_2_12___testBumpTransactionalEpoch_String__quorum_kraft/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14078/2/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBalancePartitionLeaders__/)
```

**Comment by olalamichelle** [comment]
@divijvaidya BTW I recently started contributing to Kafka, do you have any suggested new feature or bug fixing ticket for me?

**Comment by divijvaidya** [comment]
Hey @olalamichelle 
I listed some items recently here - https://lists.apache.org/thread/cl80kbn7cp4co0cxxwgw3tmqn9lkcw7v 
Please feel free to reach out to me via email (divijv@apache.org) if you still don't find anything that looks interesting to you.

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java
Comment (showuon) [inline clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java:253]: Could you add some java doc on what we're trying to do in this method? Same as below. Thanks.

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java
Comment (showuon) [inline clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java:304]: Should we log anything here for future troubleshooting?

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java
Comment (showuon) [inline clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java:238]: I can see it is basically identical to the `MockScheduler`. Could you explain why can't we use `MockScheduler` directly here?

**Review by showuon — COMMENTED** [review]
@olalamichelle , thanks for the PR. Left some comments.

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java
Comment (olalamichelle) [reply clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java:253]: Done

Comment (olalamichelle) [reply clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java:253]: This is a test-only class so I think it should be fine? We should make sure scheduled thing does not have exceptions?

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java
Comment (olalamichelle) [reply clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java:304]: This is a test-only class so I think it should be fine? We should make sure scheduled thing does not have exceptions?

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java
Comment (olalamichelle) [reply clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java:238]: Yes it is very similar but I have to create my own mock for 2 reasons:
1. MockScheduler.schedule method does not take a period parameter which schedules a periodical task. I can add this function to MockScheduler but then I will also implement the corresponding addWaiter which is pretty much just put all the MockExecutorService code into MockScheduler. I don't think it is a good idea since it will make MockScheduler very cumbersome. Moreover, if I do that its schedule interface cannot take a ExecutorService parameter and it needs to take care of the execution itself inside. See below for more details.
2. MockScheduler is really just a scheduler that does not take care of the execution. It just submits tasks to the executor service. The flakiness of the test comes from the real clock based executor service where, because of CPU scheduling, cannot be 100% accurate on timings. We need a callback-based one in the test to make sure it is reliable. It is just the logic (since there principles are the same) behind a mock scheduler and a mock executor service is the same that we have to use MockTime and be callback-based so they are very similar.

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java
Comment (showuon) [reply clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/RefreshingHttpsJwksTest.java:304]: OK, but we sometimes do `e.printStackTrace();` for this kind of situation like here:
https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java#L272

It will be helpful when this test someday become flaky. Could we add it?

**Review by showuon — COMMENTED** [review]
LGTM! Just one minor comment.

**Review by showuon — APPROVED** [review]
LGTM!

**Comment by showuon** [comment]
Let's wait for the CI build results and also see if @divijvaidya wants to have another review.

**Review by divijvaidya — APPROVED** [review]
Looks good!

**Comment by showuon** [comment]
Failed tests are unrelated.

**Comment by olalamichelle** [comment]
Thank you, Luke!

**Pull request merged**
